app/src/main/kotlin/compiler/intermediate/ControlFlow.kt

```diff
@@ -363,20 +405,94 @@ object ControlFlow {
+                    val elementAddress =
+                        IFTNode.Add(array, IFTNode.Multiply(index, IFTNode.Const(memoryUnitSize.toLong())))
+                    val reg = Register()
+                    cfgBuilder.addNextTree(
+                        IFTNode.RegisterWrite(
+                            reg,
+                            IFTNode.MemoryRead(elementAddress)
+                        )
+                    )
+                    expressionTypes[Ref(astNode.expression)]?.let {
+                        if ((it as Type.Array).elementType is Type.Array) {
+                            cfgBuilder.addNextCFG(arrayMemoryManagement.genRefCountIncrement(IFTNode.MemoryRead(elementAddress)))
+                        }
+                    }
+                    cfgBuilder.addNextCFG(arrayMemoryManagement.genRefCountDecrement(array, expressionTypes[Ref(astNode.expression)]!!))
+                    IFTNode.RegisterRead(reg)
+                }
+                is Expression.ArrayAllocation -> {
+                    val size = makeCFGForSubtree(astNode.size)
+                    val expressions = astNode.initialization.map { makeCFGForSubtree(it) }
+                    val arrayType = expressionTypes[Ref(astNode)]!! as Type.Array
+                    val allocation = arrayMemoryManagement.genAllocation(size, expressions, arrayType, astNode.initializationType)
+                    cfgBuilder.addNextCFG(allocation.first)
+                    if (arrayType.elementType is Type.Array) {
+                        expressions.forEach {
+                            cfgBuilder.addNextCFG(arrayMemoryManagement.genRefCountDecrement(it, arrayType.elementType))
+                        }
+                    }
+                    allocation.second
+                }
             }
         }
 
+        val (targetArray, index) = if (target is AssignmentTarget.ArrayElementTarget)
+            Pair(makeCFGForSubtree(target.element.expression), makeCFGForSubtree(target.element.index))
+        else Pair(null, null)
         val result = makeCFGForSubtree(expression)
 
         // build last tree into CFG, possibly wrapped in variable write operation
-        if (targetVariable != null) {
-            val owner = variableProperties[Ref(targetVariable)]!!.owner
-            cfgBuilder.addNextTree(variableAccessGenerators[Ref(owner)]!!.genWrite(targetVariable, result, owner === currentFunction))
-        } else if (accessNodeConsumer == null) {
-            cfgBuilder.addNextTree(result)
+        if (target != null) {
+            when (target) {
+                is AssignmentTarget.VariableTarget -> {
+                    val owner = variableProperties[Ref(target.variable)]!!.owner
+                    cfgBuilder.addNextTree(variableAccessGenerators[Ref(owner)]!!.genWrite(target.variable, result, owner === currentFunction))
+                }
+                is AssignmentTarget.ArrayElementTarget -> {
```

Comment:
Why aren't we increasing the refcount here? The assigned variable will be a new reference to the array, right?

Reply:
The same reason as above